### ISSUE TITLE
Fixed amount cart rule is duplicated after order splitting

### ISSUE THREAD

**Issue by mcaldex** [PrestaShop/PrestaShop#36177]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When the customer adds a fixed amount cart rule to his cart, and order splitting happens (when two different products are shipped only by different carriers), the cart rule is applied to both orders.

### Expected behavior

Apply the cart only only to one of the orders (first one preferably).
But maybe the order splitting feature may be revisited completely.

### Steps to reproduce

1. create two carriers in backoffice
2. create to products in backoffice
3. associate the first carrier to the first product, then associate the second carrier to the second product:
![image](https://github.com/PrestaShop/PrestaShop/assets/79152536/7007c7b8-085c-45cc-a69b-89403d610db5)
![image](https://github.com/PrestaShop/PrestaShop/assets/79152536/e42336fd-65dd-4dfa-9b54-3a64347a454b)
4. create a fixed amount cart rule:
![image](https://github.com/PrestaShop/PrestaShop/assets/79152536/ba65fc68-3080-4911-acf2-54ceaf2a47a6)
5. add the products to the cart in frontoffice, and then add the cart rule, see the calculations are correct:
![image](https://github.com/PrestaShop/PrestaShop/assets/79152536/0dda10f1-9842-4e3c-a1cc-ae6b087f9f78)
6. finalise the checkout process and validate the order
7. see the cart is splitted and the cart rule has been duplicated on both orders:
![image](https://github.com/PrestaShop/PrestaShop/assets/79152536/5c893786-fce0-4419-8e1f-e677de60ae08)
![image](https://github.com/PrestaShop/PrestaShop/assets/79152536/2ad6771f-3417-4bbf-976e-600efa35a276)

### PrestaShop version(s) where the bug happened

8.1.1

### PHP version(s) where the bug happened

8.1

### Your company or customer's name goes here (if applicable).

Hostinato

**Comment by mcaldex** [comment]
Related epic bug thread: #15491

**Comment by AureRita** [comment]
Hi @MHostinato ,

Thank you for your report, I reproduce the issue with PrestaShop version 8.1.x, as you can see : 

[recording(209).webm](https://github.com/PrestaShop/PrestaShop/assets/100699445/de269647-664f-4115-b0d1-0906679318fc)


Thank you for this clear issue ! I'll add this to the backlog so it can be fixed.

Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you. 

If you fix the issue on your end, please [contribute it back to the project](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thanks

**Comment by jouino16** [comment]
In my humble opinion, I think this issue is duplicate of this one https://github.com/PrestaShop/PrestaShop/issues/11622.

I tackled this issue while working on this module [Multi Shipping / Multiple Carriers Order](https://addons.prestashop.com/en/shipping-carriers/93768-multi-shipping-multiple-carriers-order.html) (also available [here](https://smartunicorn.fr/produit/module-prestashop-commande-avec-transporteurs-livraisons-multiples/) - check for better FAQ section readability).

In a nutshell, this module enables multi shipping / multi carrier orders in PrestaShop. The order will be split based on the number of the carriers. Here is [demo video](https://www.youtube.com/watch?v=5QQMzO3-EdI&ab_channel=SmartUnicorn).

Regarding how I handled discount codes, please check the module documentation [Multi Shipping - Multiple Carriers Order_documentation.pdf](https://github.com/user-attachments/files/19740632/Multi.Shipping.-.Multiple.Carriers.Order_documentation.pdf) starting from page 37 (section "4.2. Multi carrier order with discount code"). You will see screenshots on how it is handled in front office, back office, emails and invoices. Also, please pay attention to the related limitations in section "5. Limitations and unsupported features" (page 60).

Hope this helps!

**Comment by boo-code** [comment]
Reproduced on 9.2.x and traced. Posting the mechanism before touching it, because the fix lands on the money fields and the shape is a decision rather than a detail.

Reproduction is a Behat scenario in the `order` suite: two products sharing no carrier, a 20.00 fixed voucher, order placed through `AddOrderFromBackOfficeCommand`, which reaches `PaymentModule::validateOrder()` like the front office does. Measured:

```
orders produced by cart: #6 total_paid 89.80, #7 total_paid 94.80  -> 184.60
cart total, Cart::BOTH:                                               205.80
discount recorded, tax excl: 20.00 + 20.00                         -> 40.00
```

So the gap is 21.20, exactly one extra tax-included application of the 20.00 voucher, and it shows up in `total_paid`, not only in the discount lines. That is also what #11622 reports as "Order Amount is wrong".

There are two computation sites, and only the second is the one people find:

- `PaymentModule.php:1122` / `:1126` set `total_discounts_tax_excl` / `_tax_incl` from `$cart->getOrderTotal(..., Cart::ONLY_DISCOUNTS, $order->product_list, $carrierId)`, and `:1163` sets `total_paid` from the same call with `Cart::BOTH`. These run inside `createOrderFromCart()`, once per package.
- `PaymentModule::createOrderCartRules()` (`:1236`, called at `:557`) then writes the `order_cart_rule` rows, again once per order.

Both give the calculator only that package's rows, and `CartRuleCalculator` caps an amount at the rows it is handed:

```php
$discountConverted = min($discountConverted, $cartRule->reduction_tax ? $totalTaxIncl : $totalTaxExcl);
```

Each package is worth far more than the voucher, so each package gets the whole of it. Nothing in either site knows the cart is being shared out.

The code already knows the split case is different, in one place only. At `:1283` the partial-use remainder voucher is gated on `count($order_list) == 1`:

```php
if (count($order_list) == 1 && $remainingValue > 0 && $cartRule->partial_use == 1 && ...)
```

so whoever wrote that saw that a split cart needs different handling for an amount discount, and guarded the remainder path without guarding the duplication.

One more measurement that may matter more than the rest: running the same cart with `improved_shipment` enabled produces a single order, the totals add up to the cart total, and the discount is recorded once. So this is a legacy-path defect that the new shipment model does not inherit.

I have a patch for `createOrderCartRules()` that caps each order's value by what the sibling orders of the same cart already received, and it works at that level, but on its own it is a half fix: `total_discounts` and `total_paid` are written earlier in `createOrderFromCart()` and never consult `order_cart_rule`, so the order rows and the order totals would disagree. Correcting the totals means changing what a split order says it was paid, which is why I stopped here.

Two shapes, and they differ for the customer:

1. Cap with spillover. The first order absorbs as much of the amount as its own total allows, the next takes the remainder. Preserves the sum, never exceeds any single order's total, and matches the semantics `partial_use` already has within one cart.
2. Proportional distribution. Each order gets the share of the amount that its products represent. Also preserves the sum, and every invoice reflects its own weight, but each invoice then shows a fraction of the voucher the customer was quoted.

The reporter here asked for the whole thing on one order, which is shape 1 without the spillover. Which do you want? I would rather have that answer than guess on fields that decide what a shop bills.

The reproduction is written and red; I will attach it to whichever shape is chosen.
